include/cuco/static_set.cuh

```diff
@@ -533,6 +534,108 @@ class static_set {
                   OutputIt output_begin,
                   cuda_stream_ref stream = {}) const;
 
+  /**
+   * @brief Retrieves the matched key in the set corresponding to all probe keys in the range
+   * `[first, last)`
+   *
+   * If key `k = *(first + i)` has a match `m` in the set, copies a `cuco::pair{k, m}` to
+   * unspecified locations in `[output_begin, output_end)`. Else, does nothing.
+   *
+   * @note This function synchronizes the given stream. For asynchronous execution use
+   * `retrieve_async`.
+   * @note Behavior is undefined if the size of the output range exceeds
+   * `std::distance(output_begin, output_end)`.
+   * @note Behavior is undefined if the given key has multiple matches in the set.
+   *
+   * @tparam InputIt Device accessible input iterator
+   * @tparam OutputIt1 Device accessible output iterator whose `value_type` can be constructed from
+   * `ProbeKey`
+   * @tparam OutputIt2 Device accessible output iterator whose `value_type` can be constructed from
+   * `Key`
+   *
+   * @param first Beginning of the sequence of probe keys
+   * @param last End of the sequence of probe keys
+   * @param output_probe Beginning of the sequence of the probe keys that have a match
+   * @param output_match Beginning of the sequence of the matched keys
+   * @param stream CUDA stream used for retrieve
+   *
+   * @return The iterator indicating the last valid pair in the output
+   */
+  template <typename InputIt, typename OutputIt1, typename OutputIt2>
+  cuda::std::pair<OutputIt1, OutputIt2> retrieve(InputIt first,
+                                                 InputIt last,
+                                                 OutputIt1 output_probe,
+                                                 OutputIt2 output_match,
+                                                 cuda_stream_ref stream = {}) const;
+
+  /**
+   * @brief Asynchronously retrieves the matched key in the set corresponding to all probe keys in
+   * the range `[first, last)`
+   *
+   * If key `k = *(first + i)` has a match `m` in the set, copies a `cuco::pair{k, m}` to
+   * unspecified locations in `[output_begin, output_end)`. Else, does nothing.
+   *
+   * @note Behavior is undefined if the size of the output range exceeds
+   * `std::distance(output_begin, output_end)`.
+   * @note Behavior is undefined if the given key has multiple matches in the set.
```

Comment:
Is it undefined or do we return the first matching occurrence of the key?

Reply:
It's the first element for scalar probing but undefined behavior for CG-based algorithms so undefined behavior is accurate.